snuba/datasets/discover.py

```diff
@@ -55,6 +61,67 @@
 logger = logging.getLogger(__name__)
 
 
+EVENT_CONDITION = FunctionCallMatch(
+    None,
+    Param("function", Or([StringMatch(op) for op in BINARY_OPERATORS])),
+    (
+        ColumnMatch(None, None, StringMatch("type")),
+        Param("event_type", AnyExpression()),
+    ),
+)
+
+
+def match_query_to_table(
+    query: Query, events_only_columns: ColumnSet, transactions_only_columns: ColumnSet
+) -> (str, bool, bool):
+    condition = query.get_condition_from_ast()
+    condition_columns = set()
+    event_types = []
+    if condition:
+        for cond in condition:
+            if Any(Column).match(cond):
+                condition_columns.add(cond)
+
+            result = EVENT_CONDITION.match(cond)
+            if not result:
+                continue
+
+            event_type_param = result.expression("event_type")
+            event_type = None
+            if isinstance(event_type_param, Column):
+                event_type = result.expression("event_type").column_name
+            else:
+                event_type = result.expression("event_type").value
+            if result:
+                if result.string("function") == ConditionFunctions.EQ:
+                    event_types.append(event_type)
+                elif result.string("function") == ConditionFunctions.NEQ:
+                    if event_type == "transaction":
+                        return EVENTS
+
+    if len(event_types) == 1 and event_types[0] == "transaction":
+        return TRANSACTIONS
+    elif len(event_types) >= 1:
+        return EVENTS
+
+    if any(events_only_columns.get(col.column_name) for col in condition_columns):
+        return EVENTS
+    if any(transactions_only_columns.get(col.column_name) for col in condition_columns):
+        return TRANSACTIONS
+
+    has_transaction_columns = False
+    for col in query.get_all_ast_referenced_columns():
+        if events_only_columns.get(col.column_name):
+            return EVENTS
+        elif transactions_only_columns.get(col.column_name):
+            has_transaction_columns = True
+
+    if has_transaction_columns:
+        return TRANSACTIONS
+
```

Comment:
Isn't this already covering the conditions between line 113 and 121 ?

Reply:
You mean between 107 and 110? Those are checking columns inside the conditions, this is checking all columns.

Reply:
Yeah, but the result of this is a superset of the result of likes 107-110.
If a columnn is in the conditions it is also within all columns, if you were returning events at line 107-110 you still be returning events by running this check and same for transactions. I am saying you do not need the conditions between 107-110

Reply:
I understand this processes the condition columns twice, but we do want to prefer columns in the condition over other columns, again because those columns were explicitly added by the user. I'm not as sold on this concept as preferring if the user has specified an event type, this was preserved from the old behaviour.